### PR TITLE
Added in place geometry if geom isn't filled

### PR DESCRIPTION
**Pull request by alexgleith**
Hey, I noticed that some tweets don’t have geometry even though they do have a location added, so I’ve got a fix here that looks for a ‘place’.

### CODE REVIEW

**Comment by edsu** [comment]
Interesting, I haven't noticed that before. I'd love to see an example tweet if you have one.

I wonder, rather than guessing at a point where the tweet came from in the bounding box, why don't you [represent the bounding box](http://geojson.org/geojson-spec.html#bounding-boxes) itself in geojson?

**Comment by alexgleith** [comment]
Hi @edsu 

Here's an example:

`{"contributors": null, "truncated": false, "text": "@alexgleith @kireland_siba @philspatial It could possibly be setup using the #locate16 tag", "is_quote_status": false, "in_reply_to_status_id": 717560975633678337, "id": 717568921981702144, "favorite_count": 0, "entities": {"symbols": [], "user_mentions": [{"id": 48219656, "indices": [0, 11], "id_str": "48219656", "screen_name": "alexgleith", "name": "alex leith"}, {"id": 308316454, "indices": [12, 26], "id_str": "308316454", "screen_name": "kireland_siba", "name": "SIBA"}, {"id": 377282871, "indices": [27, 39], "id_str": "377282871", "screen_name": "philspatial", "name": "Phil Delaney"}], "hashtags": [{"indices": [77, 86], "text": "locate16"}], "urls": []}, "retweeted": false, "coordinates": null, "source": "<a href=\"http://twitter.com\" rel=\"nofollow\">Twitter Web Client</a>", "in_reply_to_screen_name": "alexgleith", "in_reply_to_user_id": 48219656, "retweet_count": 0, "id_str": "717568921981702144", "favorited": false, "user": {"follow_request_sent": false, "has_extended_profile": false, "profile_use_background_image": false, "default_profile_image": false, "id": 19001296, "profile_background_image_url_https": "https://abs.twimg.com/images/themes/theme14/bg.gif", "verified": false, "profile_text_color": "333333", "profile_image_url_https": "https://pbs.twimg.com/profile_images/344981663/pinOrange_normal.png", "profile_sidebar_fill_color": "EFEFEF", "entities": {"url": {"urls": [{"url": "https://t.co/y4m7FP0zFU", "indices": [0, 23], "expanded_url": "https://www.linkedin.com/in/hughsaalmans", "display_url": "linkedin.com/in/hughsaalmans"}]}, "description": {"urls": []}}, "followers_count": 446, "profile_sidebar_border_color": "000000", "id_str": "19001296", "profile_background_color": "131516", "listed_count": 68, "is_translation_enabled": false, "utc_offset": 36000, "statuses_count": 2614, "description": "I wonder what Latitude or Longitude I've got to? (Alice had no idea what Latitude was, or Longitude either, but thought they were nice grand words to say.)", "friends_count": 255, "location": "Sydney, AU", "profile_link_color": "FA743E", "profile_image_url": "http://pbs.twimg.com/profile_images/344981663/pinOrange_normal.png", "following": true, "geo_enabled": true, "profile_banner_url": "https://pbs.twimg.com/profile_banners/19001296/1446340035", "profile_background_image_url": "http://abs.twimg.com/images/themes/theme14/bg.gif", "screen_name": "minus34", "lang": "en", "profile_background_tile": true, "favourites_count": 952, "name": "Hugh Saalmans", "notifications": false, "url": "https://t.co/y4m7FP0zFU", "created_at": "Wed Jan 14 23:21:52 +0000 2009", "contributors_enabled": false, "time_zone": "Sydney", "protected": false, "default_profile": false, "is_translator": false}, "geo": null, "in_reply_to_user_id_str": "48219656", "lang": "en", "created_at": "Wed Apr 06 04:25:37 +0000 2016", "in_reply_to_status_id_str": "717560975633678337", "place": {"full_name": "Sydney, New South Wales", "url": "https://api.twitter.com/1.1/geo/id/0073b76548e5984f.json", "country": "Australia", "place_type": "city", "bounding_box": {"type": "Polygon", "coordinates": [[[150.520928608, -34.1183470085], [151.343020992, -34.1183470085], [151.343020992, -33.578140996], [150.520928608, -33.578140996]]]}, "contained_within": [], "country_code": "AU", "attributes": {}, "id": "0073b76548e5984f", "name": "Sydney"}, "metadata": {"iso_language_code": "en", "result_type": "recent"}}
`
This was acquired using this query a few days ago `python twarc.py "#locate16" >> tweets.json`. There are a few others like this.

I didn't want to grab the bounding box simply because keeping them all points is useful for my use-case, but you might be right in that capturing either type could be good, and handling the different use-cases client side would be better. It's a trivial change and I'd be happy to make it if you'd then accepr the PR!

**Comment by edsu** [comment]
That's really interesting that a `Place` can be attached without `geo` or `coordinates` top level stanzas being set. Thanks for sharing that!

I would prefer it if the default behavior was to represent the bounding box, since that's what is in the data. But I'm open to you adding a command line option for converting the bounding box to a point if that is useful.

**Comment by alexgleith** [comment]
Ok, I've updated the PR with command line options and bounding box as a default. I also added fuzzing (adding noise) to the centroids, as I did this for my use-case and figured it might be useful. It still clusters around the centroid rather than all throughout the polygon, but it's better than having them all clumped up together, I think. I also added doco.

File: utils/geojson.py [hunk 1]
```diff
@@ -30,6 +45,69 @@
...
+	            "type": "Feature",
+	            "geometry": {
+	                "type": "Polygon",
+	                "coordinates": [
+	                	[
+		                    bbox[0],
+		                    bbox[1],
+		                    bbox[2],
+		                    bbox[3],
+		                    bbox[0]
```
Comment (edsu) [inline utils/geojson.py]: Why is this here?

**Comment by edsu** [comment]
I had to do some cleanup and remove some duplicated code. I think I also removed an error in the way you were expressing the bounding boxe. You can see it here over on [this branch](https://github.com/edsu/twarc/tree/boundingboxes). Can you give it a try?

**Comment by alexgleith** [comment]
Hi Ed

Your cleanup is great, except for the box. It's best to store this as a
valid GeoJSON Polygon.

The Twitter bbox has four coordinates representing the bounding box, but a
valid polygon has a final coordinate back at the start (so five sets of
coordinates). It also needs to be wrapped in another set of square
brackets. See here: http://geojsonlint.com/

So I'll update this and test it and add it to my PR, if you don't mind?

Regards,

Alex
On Mon, 11 Apr 2016 at 12:21 AM, Ed Summers <notifications@github.com>
wrote:

> I had to do some cleanup and remove some duplicated code. I think I also
> removed an error in the way you were expressing the bounding boxe. You can
> see it here over on this branch
> <https://github.com/edsu/twarc/tree/boundingboxes>. Can you give it a try?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/edsu/twarc/pull/92#issuecomment-207992698>
>
-- 
Alex Leith
0419 189 050

**Comment by alexgleith** [comment]
Ah sorry, I didn't see this.
On Mon, 11 Apr 2016 at 7:36 AM, Ed Summers <notifications@github.com> wrote:

> Merged #92 <https://github.com/edsu/twarc/pull/92>.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/edsu/twarc/pull/92#event-621685027>
>
-- 
Alex Leith
0419 189 050

**Comment by alexgleith** [comment]
Hey @edsu, I've reverted the storage of the BBOX so that it's back to the way I had it originally. https://github.com/edsu/twarc/pull/95

This also fixes an issue where the script was no longer only writing tweets with location info attached.

Here's an example of output feature:

`{
  "type": "FeatureCollection", 
  "features": [
    {
      "geometry": {
        "type": "Polygon", 
        "coordinates": [
          [
            [
              147.133660416, 
              -43.014122643
            ], 
            [
              147.613383232, 
              -43.014122643
            ], 
            [
              147.613383232, 
              -42.655375527
            ], 
            [
              147.133660416, 
              -42.655375527
            ], 
            [
              147.133660416, 
              -43.014122643
            ]
          ]
        ]
      }, 
      "type": "Feature", 
      "properties": {
        "name": "alex leith", 
        "url": "http://twitter.com/alexgleith/status/717566966932766721", 
        "text": "This is a test #locate16.", 
        "created_at": "Wed Apr 06 04:17:51 +0000 2016", 
        "profile_image_url": "http://pbs.twimg.com/profile_images/3540938082/03ae11e7b0afe008984e633f9510696d_normal.jpeg", 
        "screen_name": "alexgleith"
      }
    }
  ]
}`
